# Building a decision tree for regression


In this notebook, we will consider an example to find the best feature to build a decision tree for a regression problem. We again will consider the golf palyer dataset, the target feature is the number of players that played.


To build a decision tree regression, the feature that gives has the highest standard deviation reduction will be considerd.

We will start our analysis by considering the Outlook feature. 

In [1]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_excel("golf_players.xlsx")

In [4]:
print(df)

    Day   Outlook  Temp Humidity    Wind  Golf Players
0     1     Sunny   Hot     High    Weak            25
1     2     Sunny   Hot     High  Strong            30
2     3  Overcast   Hot     High    Weak            46
3     4      Rain  Mild     High    Weak            45
4     5      Rain  Cool   Normal    Weak            52
5     6      Rain  Cool   Normal  Strong            23
6     7  Overcast  Cool   Normal  Strong            43
7     8     Sunny  Mild     High    Weak            35
8     9     Sunny  Cool   Normal    Weak            38
9    10      Rain  Mild   Normal    Weak            46
10   11     Sunny  Mild   Normal  Strong            48
11   12  Overcast  Mild     High  Strong            52
12   13  Overcast   Hot   Normal    Weak            44
13   14      Rain  Mild     High  Strong            30


In [7]:
mean_players = df["Golf Players"].mean()
std_players = df["Golf Players"].std()

print(mean_players)
print(std_players)

39.785714285714285
9.672948571450904


### The global standard deviation for players 

The global standard deviation for players is 9.67.

### Outlook 

We will find the mean and standard deviations for outlook Sunny, outlook Overcast and outlook Rain.

#### Sunny Outlook

In [8]:
df_sunny = df[df["Outlook"]=="Sunny"]
print(df_sunny)

    Day Outlook  Temp Humidity    Wind  Golf Players
0     1   Sunny   Hot     High    Weak            25
1     2   Sunny   Hot     High  Strong            30
7     8   Sunny  Mild     High    Weak            35
8     9   Sunny  Cool   Normal    Weak            38
10   11   Sunny  Mild   Normal  Strong            48


In [14]:
mean_df_sunny = df_sunny["Golf Players"].mean()
std_df_sunny = df_sunny["Golf Players"].std()


print("Number of Overcast observations: ", df_sunny.shape[0])

print(mean_df_sunny)
print(std_df_sunny)

Number of Overcast observations:  5
35.2
8.700574693662482


#### Overcast Outlook

In [15]:
df_overcast = df[df["Outlook"]=="Overcast"]
print(df_overcast)

    Day   Outlook  Temp Humidity    Wind  Golf Players
2     3  Overcast   Hot     High    Weak            46
6     7  Overcast  Cool   Normal  Strong            43
11   12  Overcast  Mild     High  Strong            52
12   13  Overcast   Hot   Normal    Weak            44


In [16]:
mean_df_overcast = df_overcast["Golf Players"].mean()
std_df_overcast = df_overcast["Golf Players"].std()

print("Number of Overcast observations: ", df_overcast.shape[0])

print(mean_df_overcast)
print(std_df_overcast)

Number of Overcast observations:  4
46.25
4.031128874149275


#### Rain Outlook

In [17]:
df_rain = df[df["Outlook"]=="Rain"]
print(df_rain)

    Day Outlook  Temp Humidity    Wind  Golf Players
3     4    Rain  Mild     High    Weak            45
4     5    Rain  Cool   Normal    Weak            52
5     6    Rain  Cool   Normal  Strong            23
9    10    Rain  Mild   Normal    Weak            46
13   14    Rain  Mild     High  Strong            30


In [18]:
mean_df_rain = df_rain["Golf Players"].mean()
std_df_rain = df_rain["Golf Players"].std()

print("Number of Rain observations: ", df_rain.shape[0])

print(mean_df_rain)
print(std_df_rain)

Number of Rain observations:  5
39.2
12.153188881935472


#### Summary of standard deviation for the outlook feature

Total observations = 14

Sunny 

number of observations=5, std = 8.7

Overcast 

number of observations=4, std = 4.03

Rain 
number of observations=5, std = 12.15


Weighted std for outlook = (5/14)* 8.7 + (4/14)* 4.03 + (5/14)* 12.15 

In [20]:
print("Weighted std for Outlook is: ", (5/14)* 8.7 + (4/14)* 4.03 + (5/14)* 12.15 )

Weighted std for Outlook is:  8.597857142857142


Std reduction for outlook is = Global std of players - weighted std of Outlook = 9.67 - 8.6 = 1.07

This process has to be repeated for other features. The feature with the highest Std reduction should be considered as the starting node to build the decision tree regressor. 